### ISSUE TITLE
Low training  P   R   mAP50  mAP50-95 but high Val  P R mAP50  mAP50-95

### ISSUE THREAD

**Issue by joeljohn135** [ultralytics/ultralytics#3611]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I have searched the issues but couldn't find any solutions.
I am running a custom dataset consisting of 1 class Gun which has 3893 training images, 1111 validation images and 557 test images. Locally
I have trained it used the command 
```
import os
from ultralytics import YOLO
model = YOLO("yolov8n.pt") 
model.train(data="config.yaml", epochs=100,amp=False) 
````
i had to switch off amp since its causing an error with the GPU if i use it. (Nvidia GTX 1650 16GB RAM 4GB VRAM )
While training it for first 100 epochs the metrics show very low training values as given below 
![image](https://github.com/ultralytics/ultralytics/assets/79360318/b06671a2-656a-4116-b3ed-19268e1e2c46)
![image](https://github.com/ultralytics/ultralytics/assets/79360318/2b609c9c-c68c-4dcb-b103-49e6b5bd39fb)
![results](https://github.com/ultralytics/ultralytics/assets/79360318/769621ff-7a92-4bfb-97cd-c346e35418a2)
![R_curve](https://github.com/ultralytics/ultralytics/assets/79360318/6bfe71e0-dfbb-4d5a-b9d9-6f47390e3fa8)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/79360318/21ce7813-cfe7-41b7-9650-0bce872ac93a)
![P_curve](https://github.com/ultralytics/ultralytics/assets/79360318/52272d0b-df43-4447-b3a3-35b99b353bd8)
![train_batch0](https://github.com/ultralytics/ultralytics/assets/79360318/59e3e2a5-d7b4-4fa2-9c8b-f96431879ff3)

But when i validate my model using the following command
```
!yolo task=detect mode=val model='runs/detect/train4/weights/best.pt' data=config.yaml
```
It give better results
![image](https://github.com/ultralytics/ultralytics/assets/79360318/ed080e42-4d21-44e1-9e7e-b66a160053c5)

I don't understand why my training gives poor results and validates well? 

So i ran the model again for another 100 epochs using the "last.pt" weights like this 
```
import os
from ultralytics import YOLO
model = YOLO("runs/detect/train4/weights/last.pt"") 
model.train(data="config.yaml", epochs=100,amp=False) 
```

However even after training this for 100 epochs so totally 200 the metrics doesn't seem going well the loss is reducing very slowly. 
![image](https://github.com/ultralytics/ultralytics/assets/79360318/5b18e6a5-0134-4f0e-b751-6cb9aa4b90a9)
![image](https://github.com/ultralytics/ultralytics/assets/79360318/8ee9f926-88c3-42cc-bdff-6b4262ccd754)
![P_curve](https://github.com/ultralytics/ultralytics/assets/79360318/9922ac82-5c22-4fea-a282-2ed463b04eaf)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/79360318/fd759a91-a59d-417a-ac6f-996e9d114ee8)
![R_curve](https://github.com/ultralytics/ultralytics/assets/79360318/6afb1ab7-21c8-4af5-8efc-7cdfeab73112)
![results](https://github.com/ultralytics/ultralytics/assets/79360318/4665a314-a6d5-4dd0-8881-4d24b57069ca)

Here is my config file
```
path: C:\Users\wind\Desktop\Papers\yolov8\data # dataset root dir
train: images\train  # train images (relative to 'path')
val: images\val  # val images (relative to 'path')

# Classes
names:
  0: Gun
```

I dont understand what is wrong with the training part. I have a quite big dataset and its just one class. Is it the model? is the number of epochs? in this case why does the mAP barely increase even after training for 200epochs. 

Hope you can help!

**Comment by glenn-jocher** [comment]
@joeljohn135 your results appear to indicate a lack of validation images and/or labels. Without validation labels no metrics or losses can be computed. 

If this was not intentional then it's likely your dataset is not correctly structured. See https://docs.ultralytics.com/datasets/detect/ for directions.

**Comment by glenn-jocher** [comment]
@joeljohn135 yeah I'm guessing that the training is not finding those 1111 images correctly, it's finding 0 validation images, or perhaps 0 labels for your validation images. You need to check your dataset and paths.

**Comment by joeljohn135** [comment]
Okay yes your right thank you.
It works on google colab well so it might be a path problem.
Also I had another doubt of whether to validate again after running 
```
import os
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # build a new model from scratch

# Use the model
results = model.train(data="config.yaml", epochs=100)  # train the model
```
And also is the validation metrics more important that the training one??

**Comment by glenn-jocher** [comment]
@joeljohn135 yeah, it's possible that the issue lies with the path configuration. Make sure that the paths to your dataset and labels are set correctly in your config file.

Regarding your doubt, it's generally a good practice to evaluate the model on a separate validation set after training. This allows you to assess the model's performance on unseen data and helps you to monitor if the model is overfitting or generalizing well.

Both training and validation metrics are important. Training metrics give you an idea of how well the model is learning during the training process. Validation metrics, on the other hand, give you an indication of how well the model is performing on unseen data. Both metrics together provide a comprehensive picture of the model's performance.

**Comment by glenn-jocher** [comment]
@joeljohn135 hi,

Thank you for reporting the issue. Based on the information you provided, it seems like you have already checked the paths and labels in your dataset. Since the mAP and other metrics are consistently at a low value, it is unlikely to be related to AMP (Automatic Mixed Precision). 

One possible reason for the issue could be related to the local setup of your environment. It is worth checking if all necessary dependencies and libraries are installed correctly, including CUDA, CUDNN, Python, and PyTorch. Additionally, ensure that you are using a compatible version of PyTorch for YOLOv8.

Another possibility is that there might be differences in the dataset or data preprocessing between your local setup and Google Colab. Ensure that the data and labels are processed in the same way, and the dataset structure is intact.

Finally, it could be beneficial to compare the training configuration and settings between your local setup and Google Colab. Ensure that all the training parameters, such as learning rate, number of epochs, and batch size, are set appropriately.

By analyzing the above possibilities related to the environment, dataset, and training setup, you should be able to identify the cause of the issue and resolve it.

If you have further questions or need additional assistance, please let us know.

**Comment by bit-scientist** [comment]
@joeljohn135 do you see close to zero values even after training for 10-15 epochs? By your last picture, it seems that the model just started learning.

**Comment by glenn-jocher** [comment]
Hi @bit-scientist,

From the results you provided in the screenshot, it appears that the model you are training has not converged yet. The low mAP and other metrics around 0.00xx suggest that the model is still in the early stages of learning. 

In general, training deep learning models, such as YOLOv8, requires several epochs for the model to learn and improve its performance. It's common to train for a larger number of epochs, typically in the range of 50-100 or more, to allow the model to converge and achieve better results.

I would recommend continuing the training process for more epochs and monitoring the performance periodically. You can save the model at regular intervals to keep track of its progress. Additionally, you can try adjusting the learning rate or other hyperparameters to potentially improve the training process.

Keep in mind that training deep learning models can be time-consuming, especially with complex architectures like YOLOv8. So, allowing the model to train for a sufficient number of epochs and monitoring its progress will likely help to achieve better results.

If you have further questions or need additional assistance, please let us know.

Best regards.

**Comment by sascha93rus** [comment]
Faced with the same breakdown. I use a GPU with 4GB of video memory. I set the batch = -1 parameter to automatically determine the size. The amp parameter = false, because video memory is low. During training, the batch parameter was set to 8 and about 2.5GB of video memory was used. And at the same time, the validation indicators (Box(P, R, mAP50, mAP50-95) always remained equal to 0 during training.
But I found a solution. it was necessary to set batch = 4 and everything worked. I think this problem needs to be fixed.

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for sharing your experience and the solution you found! It's great to hear that adjusting the batch size to 4 helped resolve the issue with the validation metrics remaining at zero. 🎉

It seems that the smaller batch size might be more suitable for your GPU's memory capacity, allowing the model to handle the validation process more effectively. We appreciate your suggestion and will consider looking into how the automatic batch size determination can be optimized for different hardware configurations.

If you have any more insights or questions as you continue your work, please feel free to share. Happy training!
